src/daemon/daemon.cpp

```diff
@@ -2653,6 +2664,145 @@ catch (const std::exception& e)
+                                                           *config->logger,
+                                                           server};
+
+        const auto& source_name = request->source_name();
+        const auto [_, status] = find_instance_and_react(operative_instances,
+                                                         deleted_instances,
+                                                         source_name,
+                                                         require_operative_instances_reaction);
+        if (status.ok())
+        {
+            auto generate_destination_name = [this](const CloneRequest& request) -> std::string {
+                auto is_name_already_used = [this](const std::string& destination_name) -> bool {
+                    return operative_instances.find(destination_name) != operative_instances.end() ||
+                           deleted_instances.find(destination_name) != deleted_instances.end() ||
+                           delayed_shutdown_instances.find(destination_name) != delayed_shutdown_instances.end();
+                };
+
+                if (request.has_destination_name())
+                {
+                    if (!mp::utils::valid_hostname(request.destination_name()))
+                    {
+                        throw std::runtime_error("Invalid destination virtual machine instance name: " +
+                                                 request.destination_name());
+                    }
+
+                    if (is_name_already_used(request.destination_name()))
+                    {
+                        throw std::runtime_error(request.destination_name() +
+                                                 " already exists, please choose a new name. ");
+                    }
+
+                    return request.destination_name();
+                }
+                else
+                {
+                    const std::string& source_name = request.source_name();
+                    const std::string destination_name =
+                        generate_next_clone_name(vm_instance_specs[source_name].clone_count, source_name);
+
+                    if (is_name_already_used(destination_name))
+                    {
+                        throw std::runtime_error("auto-generated name " + destination_name +
+                                                 " already exists, please specify a new name manually. ");
+                    }
+
+                    return destination_name;
+                }
+            };
+
+            const std::string destination_name = generate_destination_name(*request);
+
+            // if any of the below code throw, then roll back and clean up added spec and image record items
+            auto rollback_remove_spec_image_record_item =
+                sg::make_scope_guard([this, destination_name]() noexcept -> void {
+                    vm_instance_specs.erase(destination_name);
```

Comment:
ok, I see. It also removes the instance directory an instance. So that also means the rollback_delete_instance_folder scopeguard in `QemuVirtualMachineFactory::create_vm_and_clone_instance_dir_data` is no longer needed.